#### Individual Class Project

## Predicting Major League Baseball Game Outcomes
***
### Prepare Major League Baseball Data from baseball-reference.com

- All data is acquired from batting game logs for all 30 Major League Baseball teams for the year 2021
- All batting logs were combined into one .csv file
- Data source is baseball-reference.com

## Prepare

### Clean up and prepare data obtained to use for exploration and modeling


In [1]:
import warnings
warnings.filterwarnings('ignore')

import pandas as pd
import numpy as np

import acquire

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 4858 entries, 0 to 4857
Data columns (total 33 columns):
 #   Column                Non-Null Count  Dtype  
---  ------                --------------  -----  
 0   Team                  4858 non-null   object 
 1   Rk                    4858 non-null   int64  
 2   Gtm                   4858 non-null   int64  
 3   Date                  4858 non-null   object 
 4   Unnamed: 4            2429 non-null   object 
 5   Opp                   4858 non-null   object 
 6   Rslt                  4858 non-null   object 
 7   PA                    4858 non-null   int64  
 8   AB                    4858 non-null   int64  
 9   R                     4858 non-null   int64  
 10  H                     4858 non-null   int64  
 11  2B                    4858 non-null   int64  
 12  3B                    4858 non-null   int64  
 13  HR                    4858 non-null   int64  
 14  RBI                   4858 non-null   int64  
 15  BB                   

In [2]:
df = acquire.get_batting_log_data()
df.head()

Team  Rk  Gtm   Date Unnamed: 4  Opp   Rslt  PA  AB  R  ...  SB  \
0  Houston Astros   1    1  1-Apr          @  OAK  W,8-1  43  33  8  ...   0   
1  Houston Astros   2    2  2-Apr          @  OAK  W,9-5  45  40  9  ...   0   
2  Houston Astros   3    3  3-Apr          @  OAK  W,9-1  43  38  9  ...   0   
3  Houston Astros   4    4  4-Apr          @  OAK  W,9-2  40  36  9  ...   0   
4  Houston Astros   5    5  5-Apr          @  LAA  L,6-7  44  39  6  ...   0   

   CS     BA    OBP    SLG    OPS  LOB   #  Thr  Opp. Starter (GmeSc)  
0   0  0.273  0.395  0.545  0.941    8  11    R         C.Bassitt(49)  
1   0  0.315  0.409  0.575  0.984    9   9    L         J.Luzardo(38)  
2   0  0.324  0.405  0.550  0.954    7  10    L           C.Irvin(34)  
3   1  0.320  0.398  0.565  0.962    4  10    L          S.Manaea(33)  
4   0  0.312  0.391  0.527  0.918   11  11    L        J.Quintana(36)  

[5 rows x 33 columns]

In [3]:
df.shape

(4858, 33)

In [5]:
df.describe()

Rk          Gtm           PA           AB            R  \
count  4858.000000  4858.000000  4858.000000  4858.000000  4858.000000   
mean     81.466859    81.466859    37.426513    33.334911     4.530671   
std      46.750379    46.750379     4.970414     4.051859     3.184529   
min       1.000000     1.000000    21.000000    17.000000     0.000000   
25%      41.000000    41.000000    34.000000    31.000000     2.000000   
50%      81.000000    81.000000    37.000000    33.000000     4.000000   
75%     122.000000   122.000000    40.000000    36.000000     6.000000   
max     162.000000   162.000000    63.000000    57.000000    24.000000   

                 H           2B           3B           HR          RBI  ...  \
count  4858.000000  4858.000000  4858.000000  4858.000000  4858.000000  ...   
mean      8.127007     1.618567     0.137917     1.223549     4.321326  ...   
std       3.324612     1.340229     0.375708     1.182470     3.103705  ...   
min       0.000000     0.000000     0.000000     0.000000     0.000000  ...   
25%       6.000000     1.000000     0.000000     0.000000     2.000000  ...   
50%       8.000000     1.000000     0.000000     1.000000     4.000000  ...   
75%      10.000000     2.000000     0.000000     2.000000     6.000000  ...   
max      22.000000     9.000000     2.000000     8.000000    23.000000  ...   

               ROE          GDP           SB           CS           BA  \
count  4858.000000  4858.000000  4858.000000  4858.000000  4858.000000   
mean      0.258748     0.685056     0.455537     0.146357     0.238875   
std       0.521664     0.810693     0.782885     0.382051     0.018086   
min       0.000000     0.000000     0.000000     0.000000     0.069000   
25%       0.000000     0.000000     0.000000     0.000000     0.231000   
50%       0.000000     1.000000     0.000000     0.000000     0.238000   
75%       0.000000     1.000000     1.000000     0.000000     0.247000   
max       4.000000     5.000000     6.000000     3.000000     0.417000   

               OBP          SLG          OPS          LOB            #  
count  4858.000000  4858.000000  4858.000000  4858.000000  4858.000000  
mean      0.314458     0.400348     0.714809     6.579457    12.621655  
std       0.019108     0.033317     0.048638     2.581185     3.007252  
min       0.129000     0.100000     0.229000     0.000000     9.000000  
25%       0.304000     0.381000     0.687000     5.000000    10.000000  
50%       0.315000     0.399000     0.712000     6.000000    12.000000  
75%       0.324000     0.420000     0.747000     8.000000    15.000000  
max       0.522000     0.675000     1.138000    18.000000    25.000000  

[8 rows x 26 columns]

In [6]:
#Checking for nulls
df.isna().sum()

Team                       0
Rk                         0
Gtm                        0
Date                       0
Unnamed: 4              2429
Opp                        0
Rslt                       0
PA                         0
AB                         0
R                          0
H                          0
2B                         0
3B                         0
HR                         0
RBI                        0
BB                         0
IBB                        0
SO                         0
HBP                        0
SH                         0
SF                         0
ROE                        0
GDP                        0
SB                         0
CS                         0
BA                         0
OBP                        0
SLG                        0
OPS                        0
LOB                        0
#                          0
Thr                        0
Opp. Starter (GmeSc)       0
dtype: int64

- Unnamed column has 2429 columns.  Checking to see what the values hold, we find that it has an "@" when not null indicating the the game was played away, at another stadium and not the team's home stadium.  Will create a new column, 'is_away', to capture that feature.

In [7]:
df['is_away'] = np.where(df['Unnamed: 4']== '@', 1, 0)
df.head()

Team  Rk  Gtm   Date Unnamed: 4  Opp   Rslt  PA  AB  R  ...  CS  \
0  Houston Astros   1    1  1-Apr          @  OAK  W,8-1  43  33  8  ...   0   
1  Houston Astros   2    2  2-Apr          @  OAK  W,9-5  45  40  9  ...   0   
2  Houston Astros   3    3  3-Apr          @  OAK  W,9-1  43  38  9  ...   0   
3  Houston Astros   4    4  4-Apr          @  OAK  W,9-2  40  36  9  ...   1   
4  Houston Astros   5    5  5-Apr          @  LAA  L,6-7  44  39  6  ...   0   

      BA    OBP    SLG    OPS  LOB   #  Thr  Opp. Starter (GmeSc)  is_away  
0  0.273  0.395  0.545  0.941    8  11    R         C.Bassitt(49)        1  
1  0.315  0.409  0.575  0.984    9   9    L         J.Luzardo(38)        1  
2  0.324  0.405  0.550  0.954    7  10    L           C.Irvin(34)        1  
3  0.320  0.398  0.565  0.962    4  10    L          S.Manaea(33)        1  
4  0.312  0.391  0.527  0.918   11  11    L        J.Quintana(36)        1  

[5 rows x 34 columns]

- Column names are a bunch of initials.  Renaming the columns to better understand what the values represent.

In [10]:
# Rename columns for easier readability

df = df.rename(columns={"PA": "plate_app", "AB": "at_bats", "R": "runs_scored",
                  "H": "hits", "2B": "doubles", "3B": "triples", "BB": "bases_on_balls", 
                   "IBB": "intentional_bb", "SO": "strikeouts","HBP": "hit_by_pitch", 
                   "SH": "sac_hits", "SF": "sac_flies","ROE": "reached_on_error", "GDP": "double_plays",
                   "SB": "stolen_bases","CS": "caught_stealing", "BA": "batting_avg", "LOB": "left_on_base",
                   "#": "num_players_used", "Thr": "handedness_opp_pitcher"}, errors="raise")

df.head()

Team  Rk  Gtm   Date Unnamed: 4  Opp   Rslt  plate_app  at_bats  \
0  Houston Astros   1    1  1-Apr          @  OAK  W,8-1         43       33   
1  Houston Astros   2    2  2-Apr          @  OAK  W,9-5         45       40   
2  Houston Astros   3    3  3-Apr          @  OAK  W,9-1         43       38   
3  Houston Astros   4    4  4-Apr          @  OAK  W,9-2         40       36   
4  Houston Astros   5    5  5-Apr          @  LAA  L,6-7         44       39   

   runs_scored  ...  caught_stealing  batting_avg    OBP    SLG    OPS  \
0            8  ...                0        0.273  0.395  0.545  0.941   
1            9  ...                0        0.315  0.409  0.575  0.984   
2            9  ...                0        0.324  0.405  0.550  0.954   
3            9  ...                1        0.320  0.398  0.565  0.962   
4            6  ...                0        0.312  0.391  0.527  0.918   

   left_on_base  num_players_used  handedness_opp_pitcher  \
0             8                11                       R   
1             9                 9                       L   
2             7                10                       L   
3             4                10                       L   
4            11                11                       L   

   Opp. Starter (GmeSc)  is_away  
0         C.Bassitt(49)        1  
1         J.Luzardo(38)        1  
2           C.Irvin(34)        1  
3          S.Manaea(33)        1  
4        J.Quintana(36)        1  

[5 rows x 34 columns]

In [ ]:
#Create new column with result, is_win, where 1 is a win and 0 is a loss for every game

In [11]:
df['is_win'] = np.where(df.Rslt.str.startswith('W'), 1, 0)
df.head()

Team  Rk  Gtm   Date Unnamed: 4  Opp   Rslt  plate_app  at_bats  \
0  Houston Astros   1    1  1-Apr          @  OAK  W,8-1         43       33   
1  Houston Astros   2    2  2-Apr          @  OAK  W,9-5         45       40   
2  Houston Astros   3    3  3-Apr          @  OAK  W,9-1         43       38   
3  Houston Astros   4    4  4-Apr          @  OAK  W,9-2         40       36   
4  Houston Astros   5    5  5-Apr          @  LAA  L,6-7         44       39   

   runs_scored  ...  batting_avg    OBP    SLG    OPS  left_on_base  \
0            8  ...        0.273  0.395  0.545  0.941             8   
1            9  ...        0.315  0.409  0.575  0.984             9   
2            9  ...        0.324  0.405  0.550  0.954             7   
3            9  ...        0.320  0.398  0.565  0.962             4   
4            6  ...        0.312  0.391  0.527  0.918            11   

   num_players_used  handedness_opp_pitcher  Opp. Starter (GmeSc)  is_away  \
0                11                       R         C.Bassitt(49)        1   
1                 9                       L         J.Luzardo(38)        1   
2                10                       L           C.Irvin(34)        1   
3                10                       L          S.Manaea(33)        1   
4                11                       L        J.Quintana(36)        1   

   is_win  
0       1  
1       1  
2       1  
3       1  
4       0  

[5 rows x 35 columns]

In [16]:
wins = df.Rslt.str.startswith('W').sum()

In [17]:
losses = df.Rslt.str.startswith('L').sum()

In [19]:
df.shape[0]

4858

In [23]:
print("Does (Wins + Losses) = All entries in Dataframe:")

print("True" if (wins+losses == df.shape[0]) else "False")

Does (Wins + Losses) = All entries in Dataframe:
True


- All rows in dataframe were accounted for by looking for column starting with "W" or "L", no ties were recorded

In [25]:
# Drop unnecessary columns

df = df.drop(columns=['Rk', 'Gtm', 'Unnamed: 4'])
df.head()

Team   Date Unnamed: 4  Opp   Rslt  plate_app  at_bats  \
0  Houston Astros  1-Apr          @  OAK  W,8-1         43       33   
1  Houston Astros  2-Apr          @  OAK  W,9-5         45       40   
2  Houston Astros  3-Apr          @  OAK  W,9-1         43       38   
3  Houston Astros  4-Apr          @  OAK  W,9-2         40       36   
4  Houston Astros  5-Apr          @  LAA  L,6-7         44       39   

   runs_scored  hits  doubles  ...  batting_avg    OBP    SLG    OPS  \
0            8     9        3  ...        0.273  0.395  0.545  0.941   
1            9    14        4  ...        0.315  0.409  0.575  0.984   
2            9    13        3  ...        0.324  0.405  0.550  0.954   
3            9    11        2  ...        0.320  0.398  0.565  0.962   
4            6    11        1  ...        0.312  0.391  0.527  0.918   

   left_on_base  num_players_used  handedness_opp_pitcher  \
0             8                11                       R   
1             9                 9                       L   
2             7                10                       L   
3             4                10                       L   
4            11                11                       L   

   Opp. Starter (GmeSc)  is_away  is_win  
0         C.Bassitt(49)        1       1  
1         J.Luzardo(38)        1       1  
2           C.Irvin(34)        1       1  
3          S.Manaea(33)        1       1  
4        J.Quintana(36)        1       0  

[5 rows x 33 columns]

In [26]:
def prepare_data():
    
    import warnings
    warnings.filterwarnings('ignore')

    import pandas as pd
    import numpy as np

    import acquire
    
    #Acquire baseball batting data using function in acquire.py
    df = acquire.get_batting_log_data()
    
    # Create new column to capture that game was played away, not at home stadium
    df['is_away'] = np.where(df['Unnamed: 4']== '@', 1, 0)
    
    
    #Renaming columns for easier readability
    df = df.rename(columns={"PA": "plate_app", "AB": "at_bats", "R": "runs_scored",
                  "H": "hits", "2B": "doubles", "3B": "triples", "BB": "bases_on_balls", 
                   "IBB": "intentional_bb", "SO": "strikeouts","HBP": "hit_by_pitch", 
                   "SH": "sac_hits", "SF": "sac_flies","ROE": "reached_on_error", "GDP": "double_plays",
                   "SB": "stolen_bases","CS": "caught_stealing", "BA": "batting_avg", "LOB": "left_on_base",
                   "#": "num_players_used", "Thr": "handedness_opp_pitcher"}, errors="raise")
    
    #Create new column with result, is_win, where 1 is a win and 0 is a loss for every game
    df['is_win'] = np.where(df.Rslt.str.startswith('W'), 1, 0)
    
    
    # Drop unnecessary columns
    df = df.drop(columns=['Rk', 'Gtm', 'Unnamed: 4'])
    
    
    return df

### Takeaways:

- Created 'is_away' column to capture that the game was played away, not at home stadium
- Renamed columns for easier readability
- Created 'is_win' column to capture the result of the game, win or loss
- Dropped unnecessary columns, "Rk", "Gtm", "Unnamed: 4" and  because they were used on the website to show the count of the columns in the table, the game number of the season, and the "@" to indicate away game which a new column was created to capture.  
- Created function, prepare_data(), to clean up and prepare all data acquired as described above